posthog/hogql_queries/web_analytics/page_url_search_query_runner.py

```diff
@@ -0,0 +1,103 @@
+from posthog.hogql import ast
+from posthog.hogql.context import HogQLContext
+from posthog.hogql.query import execute_hogql_query
+from posthog.hogql_queries.web_analytics.web_analytics_query_runner import (
+    WebAnalyticsQueryRunner,
+)
+from posthog.schema import (
+    WebPageURLSearchQuery,
+    WebPageURLSearchQueryResponse,
+    CachedWebPageURLSearchQueryResponse,
+    PageURL,
+)
+
+PAGE_URL_SEARCH_DEFAULT_LIMIT = 100
+
+
+class PageUrlSearchQueryRunner(WebAnalyticsQueryRunner):
+    query: WebPageURLSearchQuery
+    response: WebPageURLSearchQueryResponse
+    cached_response: CachedWebPageURLSearchQueryResponse
+
+    def _get_url_column(self) -> ast.Expr:
+        current_url = ast.Call(name="toString", args=[ast.Field(chain=["properties", "$current_url"])])
+
+        if self.query.stripQueryParams:
+            return ast.Call(name="cutQueryStringAndFragment", args=[current_url])
+        else:
+            return current_url
+
+    def _get_search_condition(self) -> ast.Expr:
+        if self.query.searchTerm:
+            return ast.CompareOperation(
+                left=ast.Call(name="toString", args=[ast.Field(chain=["properties", "$current_url"])]),
+                op=ast.CompareOperationOp.ILike,
+                right=ast.Constant(value=f"%{self.query.searchTerm}%"),
+            )
+        return ast.Constant(value=True)
+
+    def to_query(self) -> ast.SelectQuery:
+        return self._get_hogql_query()
+
+    def _get_hogql_query(self) -> ast.SelectQuery:
+        with self.timings.measure("page_url_search_query"):
+            url_column = self._get_url_column()
+            search_condition = self._get_search_condition()
+            sampling_factor = self._sample_ratio
```

Comment:
@robbie-c I assume it is safe to use it here? I first implemented a custom version, but after some tests, it was working nicely on my local version. I remember you mentioning this could cause problems, so I could safely change it to a simpler approach here because, in the end, the way for calculating the sample ratios will be similar to this query.

Reply:
Sampling is great here!

The thing to remember is that sampling is by event, not by session or by person.

If you're counting the number of events with a given property, sampling works well, you can look at e.g. 10% of the total events and multiply the results by 10 and know that you're getting a pretty accurate result.

The problem comes if you're e.g. counting the number of unique users. If you look at 10% of the events, you'll see events from way more than 10% of the unique users. If you multiply that count by 10, you'll almost definitely get the wrong result.

Reply:
Oh, totally! I was wondering if there was something special about the `self.sample_ratio` (which uses `_get_or_calculate_sample_ratio`) because I did not see it being used on other parts, so I wasn't sure if using and, therefore, setting the cached property would affect the other queries somehow.

Reply:
Ah yeah that's fine